managed-ledger/src/main/java/org/apache/bookkeeper/mledger/dlog/DlogBasedEntryCacheImpl.java

```diff
@@ -0,0 +1,360 @@
+            return false;
+        }
+
+        if (size > 0) {
+            ByteBuf entryBuf = entry.getDataBuffer();
+            int readerIdx = entryBuf.readerIndex();
+            cachedData.writeBytes(entryBuf);
+            entryBuf.readerIndex(readerIdx);
+        }
+
+        PositionImpl position = (PositionImpl)entry.getPosition();
+        DlogBasedEntry cacheEntry = DlogBasedEntry.create(position, cachedData);
+        cachedData.release();
+        if (entries.put(position, cacheEntry)) {
+            manager.entryAdded(entry.getLength());
+            return true;
+        } else {
+            // entry was not inserted into cache, we need to discard it
+            cacheEntry.release();
+            return false;
+        }
+    }
+
+    @Override
+    public void invalidateEntries(final PositionImpl lastPosition) {
+        //todo reconstruct position's get func
+        final PositionImpl firstPosition = PositionImpl.get(-1, 0);
+
+        Pair<Integer, Long> removed = entries.removeRange(firstPosition, lastPosition, true);
+        int entriesRemoved = removed.first;
+        long sizeRemoved = removed.second;
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Invalidated entries up to {} - Entries removed: {} - Size removed: {}", ml.getName(),
+                    lastPosition, entriesRemoved, sizeRemoved);
+        }
+
+        manager.entriesRemoved(sizeRemoved);
+    }
+
+    @Override
+    public void invalidateAllEntries(long ledgerId) {
+        final PositionImpl firstPosition = PositionImpl.get(ledgerId, 0);
+        final PositionImpl lastPosition = PositionImpl.get(ledgerId + 1, 0);
+
+        Pair<Integer, Long> removed = entries.removeRange(firstPosition, lastPosition, false);
+        int entriesRemoved = removed.first;
+        long sizeRemoved = removed.second;
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Invalidated all entries on ledger {} - Entries removed: {} - Size removed: {}",
+                    ml.getName(), ledgerId, entriesRemoved, sizeRemoved);
+        }
+
+        manager.entriesRemoved(sizeRemoved);
+    }
+
+    @Override
+    public void asyncReadEntry(PositionImpl position, final ReadEntryCallback callback,
+                               final Object ctx) {
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Reading entry(id:{}) in asyncReadEntry of cache", ml.getName(), position.getEntryId());
+        }
+        DlogBasedEntry entry = entries.get(position);
+        if (entry != null) {
+            DlogBasedEntry cachedEntry = DlogBasedEntry.create(entry);
+            entry.release();
+            manager.mlFactoryMBean.recordCacheHit(cachedEntry.getLength());
+            callback.readEntryComplete(cachedEntry, ctx);
+        } else {
+            try{
+                AsyncLogReader logReader = FutureUtils.result(distributedLogManager.openAsyncLogReader(position.getDlsn()));
```

Comment:
please add a todo here to improve. you need some sort of log reader cache to avoid reopening log again and again.